File: designs/publication/views/incomplete.js
/* PubRefDb: Publication database web application.
   Index publication documents lacking information about:
   - Published date.
   - Published month.
   - Type of publication.
   - Journal.
   - Volume.
   - Pages.
   Published day and journal issue is not checked, and may be undefined.
*/
function(doc) {
    if (doc.entitytype !== 'publication') return;
    var pmid = null;
    for (var i in doc.xrefs) {
	var xref = doc.xrefs[i];
	if (xref.xdb.toLowerCase() === 'pubmed') {
	    pmid = xref.xkey;
	    break;
	}
    }
    if (!doc.published) {
	emit(doc._id, pmid);
	return;
    }
    var parts = doc.published.split('-');
    if (parts.length < 3 || parts[1] === '0' || parts[1] === '00' ) {
	emit(doc._id, pmid);
	return;
    }
    if (!doc.type) {
	emit(doc._id, pmid);
	return;
    }
    var journal = doc.journal;
    if (!journal) {
	emit(doc._id, pmid);
	return;
    }
    if (!journal.volume) {
	emit(doc._id, pmid);
	return;
    }
    if (!journal.pages) {
	emit(doc._id, pmid);
	return;
    }
}

File: incomplete.py
""" PubRefDb: Publication database web application.

List of publications with PubMed xref having incomplete information.
"""

from .base import *


class Incomplete(MethodMixin, GET):
    """Publications whose entries lack information about publication date,
    publication month, type of publication, journal, volume or pages."""

    outreprs = [JsonRepresentation,
                MedlineRepresentation,
                AtomRepresentation,
                PublicationsListCountHtmlRepresentation]

    def get_data_resource(self, request):
        publications = self.get_docs('publication/incomplete', None)
        self.sort_publications(publications, reverse=True)
        for publication in publications:
            self.normalize_publication(publication, request.application.get_url)
        return dict(title='Incomplete information',
                    resource='Publication incomplete',
                    publications=publications,
                    descr=self.__doc__)
